### PR TITLE
[PAN-1969] Implement RBAC

### PR DESCRIPTION
**Pull request by danut13**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

- Implement reusable RBAC contract
- Update scripts

### CODE REVIEW

**Comment by juanmanuel-tirado** [comment]
Tests are not passing. Is this a draft?

**Comment by danut13** [comment]
@juanmanuel-tirado No, it's not a draft, it's ready to be reviewed

We agreed to have a separate branch (RBAC), where we merge smaller but noncompilable branches. If I were to fix the issues, I would need to already the next features, which is not desirable

When the RBAC branch will be merged into main we would expect everything to pass.

File: src/PantosRBAC.sol [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            "PantosRBAC: caller is not the medium critical ops"
+        );
+    }
+
+    function _enforceIsSuperCriticalOps() internal view {
+        require(
+            msg.sender == _SUPER_CRITICAL_OPS,
+            "PantosRBAC: caller is not the super critical ops"
+        );
+    }
```
Comment (jacekv) [inline src/PantosRBAC.sol:63]: Why did you move that part of the code into a function instead directly into the modifier?

File: src/facets/PantosRegistryFacet.sol [hunk 2]
```diff
@@ -9,13 +9,29 @@ import {IPantosRegistry} from "../interfaces/IPantosRegistry.sol";
...
+     * @param _pauser The address of the pauser role
+     * @param _mediumCriticalOps The address of the medium critical ops role
+     * @param _superCriticalOps The address of the super critical ops role
+     */
+    constructor(
+        address _deployer,
+        address _pauser,
+        address _mediumCriticalOps,
+        address _superCriticalOps
+    ) PantosRBAC(_deployer, _pauser, _mediumCriticalOps, _superCriticalOps) {}
```
Comment (jacekv) [inline src/facets/PantosRegistryFacet.sol:33]: Why do you have a constructor in the facet?

This is supposed to be a facet, correct? Which means that it should be called in the context of the PantosHubProxy contract.

If you have the constructor here and initialize the RBAC roles here, it will be stored in the context of the PantosRegistryFacet and not PantosHubProxy.

Can you elaborate? Maybe I am overlooking something here :)

Comment (danut13) [reply src/facets/PantosRegistryFacet.sol:33]: Sure :)
To put everything into context, our goal is to have a reusable RBAC contract, which can be reused both in the context of the diamond pattern or for a normal contract (such as the forwarder and our PANDAS tokens(PAN + wrappers). 

When a contract which uses our RBAC contract is being called, we want to be agonstic to the type of the call. The contract should not care if it's a delegatecall or call, it should behave the same (so that we can reuse it for both patterns). This can be achieved by embedding the addresses of the roles in the complied contract. Then, no storage is being used, therefore it does not matter if it's a delegatecall or call.

This could have been achieved in two ways: Using constants or immutables. I opted for immutables, since I have the option to initialize them in the contructor. This will come in handy later on, when we will need to automate the whole process of deploying the contracts which are the roles, and also for end to end testing. If I were to use constants, then we would need to change the source code for every environment/CI test or any kind of automation.

File: src/PantosRBAC.sol (hunk 1 above)
Comment (danut13) [reply src/PantosRBAC.sol:63]: For reusability purpose. For example, if we would define now another modifier like whenNotPausedOrSuperCriticalOps, I would not have the option to reuse a modifier which uses this directly.

File: src/facets/PantosRegistryFacet.sol (hunk 2 above)
Comment (jacekv) [reply src/facets/PantosRegistryFacet.sol:33]: Alright, I get it and makes so far sense :) 

Some questions are coming up :) 
First:
Since you picked immutables (which makes total sense, as you already outline :)), the address get embedded into the runtime code. That would mean, that you would need to initialize the RBAC roles in other Facets, correct?

The PantosTransferFacet.sol does not require it, since there are not functions, which are only allowed to RBAC roles, but in the future, in case there are facets, which do require that, we would need to initialize it there too.

2nd:
Having those addresses embedded into the code, will not allow us to change it in the future anymore. If we have to change a RBAC role, which means that there is a new address, we will have to redeploy all contracts, which have the address set.
Is that intended?

Comment (danut13) [reply src/facets/PantosRegistryFacet.sol:33]: First:
Totally correct. You can see exactly this in the DiamondCutFacet. Now, we cannot reuse the one from the diamond library, because that one was tailored to the owner model. Therefore, I redefined a new one, which inherits from our RBAC contract, for exactly the reason you described. 

Second:
Yes, that is intended. The requirement for the RBAC model was that the roles addresses should never be changed. Well, that's actually impossible for the diamond pattern, no matter what option you choose, since you can always cut another facet which behaves as its code dictate, which can just ignore any roles (if it is in the storage), or just not initialize it (if it uses our model). But this is as close as we can get to this behaviour.

**Review by jacekv — APPROVED** [review]

**Comment by danut13** [comment]
We agreed on a different implementation. I will close this one.
